Bio/PDB/Mutate.py

```diff
@@ -0,0 +1,561 @@
+                break
+            atoms = unfold_entities(pdb_object[0], "A")
+            ns = NeighborSearch(atoms)
+            close_atoms = ns.search(rotamer_vector, 5)  # 5 Angstrom radii
+            for close_atom in close_atoms:
+                close_residue = close_atom.get_parent()
+                chain_if_close_atom = close_residue.get_parent()
+                if (
+                    close_atom.get_parent().get_id()[1] == res_num
+                    and chain_if_close_atom.get_id() == chain
+                ):
+                    # print("skipping_own_atom")
+                    continue
+
+                if skip_own_chain and chain_if_close_atom.get_id() == chain:
+                    # print("Skipping own chain")
+                    continue
+                if abs(close_residue.get_id()[1] - res_num) == 1 and is_backbone(
+                    close_atom
+                ):
+                    continue
+                dist = distance(close_atom.coord, rotamer_vector)
+                if dist > 6:
+                    continue
+                try:
+                    vdw_radi = (
+                        VW_RADII[close_atom.get_parent().get_resname()][
+                            close_atom.get_id()
+                        ]
+                        + VW_RADII[mutate_to][rotamer_atom]
+                    )
+                except KeyError:
+                    continue
+                attractive_force = (vdw_radi / dist) ** 6
+                vdw_energy += attractive_force**2 - attractive_force
+        if vdw_energy < lowest_energy:
+            lowest_energy = vdw_energy
+            best_rotamer = rotamer
+    return best_rotamer
+
+
+def mutate(
+    pdb_obj,
+    chain,
+    res_num,
+    mutate_to,
+    rotamer_lib=None,
+    mutation_type="best",
+):
+    """
+    Mutates a given residue based on the best fitting rotamers
+    """
+    Polypeptide.Polypeptide(
+        pdb_obj[0][chain]
+    ).get_phi_psi_list()  # This generates the xtra attribute for PHI and PSI angles for residues
+    try:
+        _residue = pdb_obj[0][chain][res_num]
+    except KeyError:
+        raise KeyError(f"Residue {res_num} not found in chain {chain}!")
+    for atom in list(_residue.get_atoms()):  # Create a copy to remove from
+        if not is_backbone(atom):
+            atom.parent.detach_child(atom.id)
+    phi, psi = (
+        round(np.rad2deg(y), -1) if y else 0
+        for y in [_residue.xtra[x] for x in ["PHI", "PSI"]]
+    )
+    # GET_ATR IS WRONG
+    sample_residue = read_sample_residue(mutate_to)
+    starting_points = np.asmatrix(
+        [sample_residue["N"], sample_residue["CA"], sample_residue["C"]]
+    )
+    end_points = np.asmatrix(
+        [_residue["N"].coord, _residue["CA"].coord, _residue["C"].coord]
+    )
+
+    sup = SVDSuperimposer.SVDSuperimposer()
+    sup.set(end_points, starting_points)
+    sup.run()
+    rot, tran = sup.get_rotran()
+
+    for atom, coords in sample_residue.items():
```

Comment:
Instead of doing this and delegating distances to a python function, you can make use of our KDTree class and query points near a certain xyz coordinate. You build the tree with the current structure (minus the sidechain you are mutating) and then search on the atoms of the newly added rotamer.